File: csharp/FizzBuzz/Implementation/FizzBuzzEngine.cs
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using Implementation.Util;

namespace Implementation
{
    public class FizzBuzzEngine
    {
        private static readonly IEnumerable<Rule> Rules = ListOfRules(
            new Rule(15, "Fizz-Buzz"),
            new Rule(3, "Fizz"),
            new Rule(5, "Buzz"));

        private static IEnumerable<Rule> ListOfRules(params Rule[] rules)
        {
            return rules;
        }

        public string Calculate(int number)
        {
            return Rules
                .Where(r => number.IsDivisibleBy(r.Denominator))
                .Select(r => r.Result)
                .HeadOption()
                .GetOrElse(() => number.ToString(CultureInfo.InvariantCulture));
        }
    }
}

File: scala/FizzBuzz/src/main/scala/core/FizzBuzzEngine.scala
package core

case class FizzBuzzEngine() {
  val rules = Array(
    (15, "Fizz-Buzz"),
    (3, "Fizz"),
    (5, "Buzz")
  )

  def calculate(number: Int): String = {
    val matchingRule = this.rules
      .filter { case (denominator, _) => isDivisibleBy(number, denominator) }
      .map { case (_, result) => result }
      .headOption

    matchingRule getOrElse number.toString
  }

  def isDivisibleBy(number: Int, denominator: Int): Boolean = {
    number % denominator == 0
  }
}
